ENH add np.nan option for zero_division in precision/recall/f-score

marctorsoc opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

First split of [this](https://github.com/scikit-learn/scikit-learn/pull/23183) to address zero_division=np.nan for precision, recall and F1 (and very close friends).

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.


#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

marctorsoc committed:
temp commit to checkout sklearn master

marctorsoc committed:
Merge branch 'sklearn_master' into marc_master

# Conflicts:
#	doc/whats_new/v0.21.rst
#	sklearn/metrics/classification.py
#	sklearn/metrics/tests/test_classification.py

marctorsoc committed:
- Changed whats_new to 0.22
- F-score only warns if both prec and rec are ill-defined
- new private method to simplify _prf_divide

marctorsoc committed:
flake8 warnings

marctorsoc committed:
Merge branch 'main' of https://github.com/scikit-learn/scikit-learn

# Conflicts:
#	sklearn/metrics/_classification.py
#	sklearn/metrics/tests/test_classification.py

marctorsoc committed:
Merge branch 'sklearn-main'

marctorsoc committed:
Merge branch 'sklearn-main'

marctorsoc committed:
Merge branch 'sklearn-main'

marctorsoc committed:
Merge branch 'sklearn-main'

marctorsoc committed:
only PRF related changes

marctorsoc committed:
Merge branch 'main' into zero_division_nan_prec_rec_f1

marctorsoc committed:
add PR number

glemaitre commented on doc/whats_new/v0.22.rst:
```diff
@@ -723,6 +723,10 @@ Changelog
   ill-defined metrics.
   :pr:`14900` by :user:`Marc Torrellas Socastro <marctorrellas>`.
 
+- |Feature| Added the :func:`metrics.nan_euclidean_distances` metric, which
+  calculates euclidean distances in the presence of missing values.
+  :issue:`12852` by :user:`Ashim Bhattarai <ashimb9>` and 
+  `Thomas Fan`_.
```
This look like a old thing to revert :)

glemaitre commented on doc/whats_new/v1.3.rst:
```diff
@@ -155,6 +155,15 @@ Changelog
+  metrics: :func:`precision_score`, :func:`recall_score`, :func:`f1_score`,
+  :func:`fbeta_score`, :func:`precision_recall_fscore_support`,
+  :func:`classification_report`.
+  When `zero_division=np.nan` and
+  there is a zero division, the metric is undefined and is excluded from averaging.
```
nitpick: could you fill-up the 88 characters line as much as possible?

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +145,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
         return sample_score.sum()
 
 
+def _nan_average(scores, weights):
```
I wonder if we should not be calling this one `nanaverage` and adding it in `sklearn.utils.extmath.py`. Any thoughts @thomasjpfan?

thomasjpfan replied:
I'm okay with including `nanaverage` into `skelarn.utils.extmath`. Although I'll make it private by calling it `_nanaverage`.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +145,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
         return sample_score.sum()
 
 
+def _nan_average(scores, weights):
```
```suggestion
def _nan_average(a, weights=None):
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -143,6 +145,32 @@ def _weighted_sum(sample_score, sample_weight, normalize=False):
 
 
+def _nan_average(scores, weights):
+    """
+    Wrapper to combine np.average and np.nanmean, so np.nan values are ignored
+    from the average and weights can be passed. Note that when possible,
+    we delegate to the prime methods.
+    """
```
```suggestion
    """Compute the weighted average, ignoring NaNs.
    
    Parameters
    ----------
    a : ndarray
        Array containing data to be averaged.

    weights : array-like, default=None
        An array of weights associated with the values in a. Each value in a
        contributes to the average according to its associated weight. The
        weights array can either be 1-D of the same shape as a. If `weights=None`,
        then all data in a are assumed to have a weight equal to one.

    Returns
    -------
    weighted_average : float
        The weighted average.

    Notes
    -----
    This wrapper to combine :func:`numpy.average` and :func:`numpy.nanmean`, so
    that :func:`np.nan` values are ignored from the average and weights can
    be passed. Note that when possible, we delegate to the prime methods.
    """
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2069,9 +2116,12 @@ def recall_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to `np.nan`, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.

        .. versionadded:: 1.3
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1528,13 +1577,15 @@ def precision_recall_fscore_support(
-        If set to "warn", this acts as 0, but warnings are also raised.
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to `np.nan`, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.

        .. versionadded:: 1.3
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1265,10 +1305,13 @@ def fbeta_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to `np.nan`, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.

        .. versionadded:: 1.3
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1117,10 +1145,13 @@ def f1_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to `np.nan`, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.

        .. versionadded:: 1.3
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1927,9 +1967,12 @@ def precision_score(
+
+        Notes:
+        - If set to "warn", this acts like 0, but a warning is also raised.
+        - If set to `np.nan`, such values will be excluded from the average.
```
```suggestion
        - If set to `np.nan`, such values will be excluded from the average.

        .. versionadded:: 1.3
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -2272,7 +2326,7 @@ def classification_report(
-    zero_division : "warn", 0 or 1, default="warn"
+    zero_division : {"warn", 0.0, 1.0, np.nan}, default="warn"
         Sets the value to return when there is a zero division. If set to
         "warn", this acts as 0, but warnings are also raised.
```
We should add the info about `np.nan` in the description and add the `versionadded` directive.

```suggestion
        "warn", this acts as 0, but warnings are also raised.

        .. versionadded:: 1.3
           `np.nan` option was added.
```

glemaitre reviewed (commented):
Just looking at the docstring. I will have a deeper look at the code and tests.

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1638,47 +1693,32 @@ def precision_recall_fscore_support(
+    elif beta == 0:
+        f_score = precision
     else:
+        # set to zero_division_value if denom 0
+        # OR if BOTH precision and recall are ill-defined
```
```suggestion
        # The score is defined as:
        # score = (1 + beta**2) * precision * recall / (beta**2 * precision + recall)
        # We set to `zero_division_value` if the denominator is 0 **or** if **both**
        # precision and recall are ill-defined.
```

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1638,47 +1693,32 @@ def precision_recall_fscore_support(
         denom = beta2 * precision + recall
-
-        denom[denom == 0.0] = 1  # avoid division by 0
+        mask = (denom == 0.0) | ((pred_sum + true_sum) == 0)
```
```suggestion
        mask = (denom == 0.0) | ((pred_sum + true_sum) == 0.0)
```

I really dislike that we do a pure comparison with floating point.
I am wondering if there is a downside with the following:

```python
mask = np.isclose(denom, 0) | np.isclose(pred_sum + true_sum, 0)
```

marctorsoc replied:
good call. Passing tests, so will use this

glemaitre commented on sklearn/metrics/_classification.py:
```diff
@@ -1143,7 +1174,8 @@ def f1_score(
     In such cases, by default the metric will be set to 0, as will f-score,
     and ``UndefinedMetricWarning`` will be raised. This behavior can be
-    modified with ``zero_division``.
+    modified with ``zero_division``. Note that if `zero_division` is np.nan,
+    scores being np.nan will be ignored for averaging.
```
```suggestion
    modified with ``zero_division``. Note that if `zero_division` is `np.nan`,
    scores being `np.nan` will be ignored for averaging.
```

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -179,9 +180,9 @@ def test_classification_report_output_dict_empty_input():
-            "f1-score": 0.0,
-            "precision": 0.0,
-            "recall": 0.0,
+            "f1-score": np.nan,
+            "precision": np.nan,
+            "recall": np.nan,
```
This case shows a different behaviour. IMO, this is a corner case and probably `np.nan` is more meaningful because the metric is not defined. However, we should acknowledge it the changelog as a change of behaviour.

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -724,9 +725,48 @@ def test_cohen_kappa():
     )
 
 
-def test_matthews_corrcoef_nan():
-    assert matthews_corrcoef([0], [1]) == 0.0
-    assert matthews_corrcoef([0, 0], [0, 1]) == 0.0
```
We should let this test for the moment because we did not modify anything in the MCC score.

glemaitre commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -1778,12 +1839,17 @@ def test_precision_recall_f1_no_labels(beta, average, zero_division):
     )
+    assert s is None
+
+    if np.isnan(zero_division):
```
A small comment within the `if` would be great to explain the early exit.

glemaitre reviewed (commented):
I think that I only miss a test to cover the case `recall + beta**2 * precision` would become null.

Otherwise, I am fine with the code.

marctorsoc committed:
PR comments

marctorsoc committed:
Merge branch 'main' into zero_division_nan_prec_rec_f1

marctorsoc committed:
fix validate_params for f1_score

marctorsoc commented:
> 

Hi @glemaitre , I addressed all your comments. Where would you put this test case you suggest? I've been looking at the code but there are so many tests I'm unsure a) whether it's already covered, b) where's the best place. Should I create just a new test for very specific edge case?

marctorsoc committed:
fix validate_params for f1_score, again

marctorsoc committed:
Merge branch 'main' into zero_division_nan_prec_rec_f1

marctorsoc commented:
Hi @glemaitre @thomasjpfan ,  kind reminder here

glemaitre commented:
Thanks. I will have a look in the day.

glemaitre commented:
I will take care about the conflicts in the same time.

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/marctorsoc/25531

glemaitre committed:
update param validation

glemaitre committed:
Update v0.22.rst

glemaitre committed:
Remove unecessary check for

glemaitre committed:
Merge branch 'zero_division_nan_prec_rec_f1' of github.com:marctorsoc/scikit-learn into pr/marctorsoc/25531

glemaitre committed:
nit: switch line

glemaitre committed:
add info in Changed models

glemaitre committed:
Merge remote-tracking branch 'origin/main' into pr/marctorsoc/25531

glemaitre commented:
Otherwise, I did solve the conflicts. So apart from the missing test, it is good to go on my side.

pinging @thomasjpfan in case you can dedicate time to get this one inside `main`.

marctorsoc committed:
merge

marctorsoc committed:
remove duplicated fix in change log

marctorsoc committed:
fix imports order

marctorsoc commented:
Sorry I meant @glemaitre above

glemaitre commented:
From the previous discussion, I was referring to:

> a test to cover the case recall + beta**2 * precision would become null.

marctorsoc commented:
> From the previous discussion, I was referring to:
> 
> > a test to cover the case recall + beta**2 * precision would become null.

I understand, but the only way to induce this to be null is by making one of the terms null. And as I pointed above, there are already tests covering no labels and no predictions, which would be the reasonable cases. Do you mean a test case with beta = np.nan?

glemaitre committed:
Merge branch 'main' into zero_division_nan_prec_rec_f1

glemaitre committed:
fix parameter validation

glemaitre commented:
Indeed, the tests are enough. I push a change that should solve the last issue.
Once the CI is green, we can merge.

marctorsoc committed:
Merge branch 'main' into zero_division_nan_prec_rec_f1

glemaitre commented:
Thanks @marctorsoc
